types/wlr_pointer_constraints_v1.c

```diff
@@ -0,0 +1,423 @@
+	constraint->accepted = true;
+	pointer_constraint_check_no_region_and_activate(constraint);
+}
+
+void wlr_pointer_constraint_v1_decline(struct wlr_pointer_constraint_v1 *constraint) {
+	wlr_log(L_DEBUG, "constraint %p declined", constraint);
+	constraint->accepted = false;
+	send_unconstrained(constraint);
+}
+
+static void resource_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static void pointer_constraint_destroy(struct wlr_pointer_constraint_v1 *constraint) {
+	if (constraint == NULL) {
+		return;
+	}
+
+	wlr_log(L_DEBUG, "destroying constraint %p", constraint);
+
+	// If it's oneshot then pointer_constraint_destroy will be called twice
+	// which will lead to a double free
+	constraint->lifetime = ZWP_POINTER_CONSTRAINTS_V1_LIFETIME_PERSISTENT;
+	send_unconstrained(constraint);
+
+	wl_resource_set_user_data(constraint->resource, NULL);
+	wl_list_remove(&constraint->link);
+	wl_list_remove(&constraint->surface_commit.link);
+	wl_list_remove(&constraint->surface_destroy.link);
+	wl_list_remove(&constraint->focus_change.link);
+	if (constraint->pointer_motion.link.next) {
+		wlr_log(L_DEBUG, "pointer_motion listeners uninstalled");
+		wl_list_remove(&constraint->pointer_motion.link);
+		wl_list_remove(&constraint->touch_motion.link);
+	}
+	free(constraint);
+}
+
+static void pointer_constraint_destroy_resource(struct wl_resource *resource) {
+	struct wlr_pointer_constraint_v1 *constraint =
+		pointer_constraint_from_resource(resource);
+
+	pointer_constraint_destroy(constraint);
+}
+
+static void pointer_constraint_set_region(struct wl_client *client,
+		struct wl_resource *resource, struct wl_resource *region) {
+	struct wlr_pointer_constraint_v1 *constraint =
+		pointer_constraint_from_resource(resource);
+
+	constraint->pending.region = region;
+}
+
+// only valid for locked_pointer
+static void pointer_constraint_set_cursor_position_hint(struct wl_client *client,
+		struct wl_resource *resource, wl_fixed_t x, wl_fixed_t y) {
+	struct wlr_pointer_constraint_v1 *constraint =
+		pointer_constraint_from_resource(resource);
+
+	constraint->pending.cursor_hint.x = wl_fixed_to_double(x);
+	constraint->pending.cursor_hint.y = wl_fixed_to_double(y);
+	constraint->pending.cursor_hint.valid = true;
+}
+
+static void handle_focus_change(struct wl_listener *listener, void *data) {
+	struct wlr_pointer_constraint_v1 *constraint =
+		wl_container_of(listener, constraint, focus_change);
+
+	struct wlr_seat *seat = data;
+	if (seat != constraint->seat) {
+		return;
+	}
+
+	send_unconstrained(constraint);
+	if (constraint->seat->pointer_state.focused_surface == constraint->surface) {
+		wlr_log(L_DEBUG, "found constraint %p for %p", constraint, seat);
+		pointer_constraint_check_no_region_and_activate(constraint);
+	}
+}
+
+// checks whether we are within the region
+static void handle_motion(struct wlr_pointer_constraint_v1 *constraint, void *data) {
```

Comment:
TBH I'd rather move this in rootston, or just provide a helper. The protocol states pretty clearly that whether or not a constraint becomes active is compositor-dependent, and wlroots can't be opinionated.

>It is up to the compositor whether to warp the pointer or require some kind of user interaction for the lock to activate.

Reply:
I don't understand this comment, I'm clearly supporting that usecase. There's a reason I have three events for the compositor, the compositor can deny/delay/accept constraints however it wishes. It is named `request_constraint` after all, not `force_constraint`.

Reply:
You can do the warp thing by setting the focused surface in the request event. The focus_change listener will then emit the constraint_active event.

Reply:
BTW if you are referring what happens when you change regions instead, then I understand. I *can* add in a region_change event if you wish so.